### PR TITLE
First pass for Derivation class structure.

### PR DESCRIPTION
**Pull request by MCazaly**
First pass to be looked at - my MATLAB is more than a little rusty and I'm not particularly familiar with OOP in MATLAB to begin with, so I would appreciate a second pair of eyes.

### CODE REVIEW

**Review by qtl-bodc — COMMENTED** [review]
Putting my Matlab knowledge deficiencies aside, Given we have sensor, parameter, equationid combinations, it looks like a good candidate for applying and thus benefiting from the OO approach  - going forward. However, the immediate issue would be trying to integrate it into Steve's framework for  o2 processing. Though there's no impact on any current work, I'm not sure if there's any great value to merging this branch directly to 'main'. Note,  we have a 'develop' branch now.

File: bgc_derivation/derivations/Derivation.m [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        function add_data(self, name, data)
+            % Add a single data variable by name.
+            %
+            % derivation.add_data(name, data)
+            % where
+            %  name is the name of the supplied variable.
+            %  data is the underlying array for the supplied variable.
+            %
+            
+            self.variables(name) = data;
```
Comment (vidkris) [inline bgc_derivation/derivations/Derivation.m:40]: Can this be used to save the coeff that is received from  'stcoeff = getPredeploymentCoefficients(metafilepath);'

File: bgc_derivation/derivations/oxygen/DOXY_201_201_301.m [hunk 2]
```diff
@@ -0,0 +1,35 @@
+classdef DOXY_201_201_301 < DOXY_X_X_301
```
Comment (vidkris) [inline bgc_derivation/derivations/oxygen/DOXY_201_201_301.m:1]: It so happens that '201_201_301','202_201_301' have the same calculation. In that case will it lead to duplication of class?

File: bgc_derivation/derivations/oxygen/DOXY_X_X_301.m [hunk 3]
```diff
@@ -0,0 +1,10 @@
+classdef DOXY_X_X_301 < Derivation
+    % Superclass for derivations using DOXY equation 301.
+    
+    % title - s DOXY_201_201_301 vr - 1.0 author - bodc/matcaz date - 20220811
```
Comment (vidkris) [inline bgc_derivation/derivations/oxygen/DOXY_X_X_301.m:4]: sorry. Should this title be DOXY_X_X_301?

**Review by vidkris — COMMENTED** [review]
A very good class structure. Seems like lot of thought has gone in. I just added a few comments.

File: bgc_derivation/derivations/oxygen/DOXY_X_X_301.m (hunk 3 above)
Comment (MCazaly) [reply bgc_derivation/derivations/oxygen/DOXY_X_X_301.m:4]: Yes it should, I'll fix that now.

File: bgc_derivation/derivations/oxygen/DOXY_201_201_301.m (hunk 2 above)
Comment (MCazaly) [reply bgc_derivation/derivations/oxygen/DOXY_201_201_301.m:1]: In this case, we would create `DOXY_X_201_301`, which perhaps I should have already done!

File: bgc_derivation/derivations/Derivation.m (hunk 1 above)
Comment (MCazaly) [reply bgc_derivation/derivations/Derivation.m:40]: I think we've resolved this with our external conversation.

File: bgc_derivation/derivations/oxygen/DOXY_X_X_301.m (hunk 3 above)
Comment (MCazaly) [reply bgc_derivation/derivations/oxygen/DOXY_X_X_301.m]: Change pushed.

File: bgc_derivation/derivations/oxygen/DOXY_201_201_301.m (hunk 2 above)
Comment (vidkris) [reply bgc_derivation/derivations/oxygen/DOXY_201_201_301.m:1]: Yes. Thankyou

**Review by vidkris — COMMENTED** [review]
Review commented incorporated. Approved

**Review by vidkris — APPROVED** [review]

**Pull request merged**